File: ACADEMIC CODES/Trie/javascript/trie_insert_search.js

function TrieNode(key) {

    this.key = key;
    this.parent = null;
    this.children = {};
    this.end = false;
}

function Trie() {
    this.root = new TrieNode(null);
}

Trie.prototype.insert = function(word) {
    var node = this.root;
    for(var i = 0; i < word.length; i++) {
        if (!node.children[word[i]]) {
            node.children[word[i]] = new TrieNode(word[i]);
            node.children[word[i]].parent = node;
        }

        node = node.children[word[i]];
        if (i == word.length-1) {
            node.end = true;
        }
    }
};

Trie.prototype.contains = function(word) {
    var node = this.root;

    for(var i = 0; i < word.length; i++) {
        if (node.children[word[i]]) {
            node = node.children[word[i]];
        } else {
            return false;
        }
    }
    return node.end;
};





var trie = new Trie();

trie.insert("cat");
trie.insert("car");
trie.insert("pick");
trie.insert("pickle");
trie.insert("dog");
trie.insert("do");

console.log(trie.contains("car"));  // true
console.log(trie.contains("picke")); // false

File: ACADEMIC CODES/Trie/python/trie_insert_search.py
class TrieNode:

    def __init__(self):
        self.children = [None]*26
        self.isEndOfWord = False
class Trie:
    def __init__(self):
        self.root = self.getNode()
    
    def getNode(self):
        return TrieNode()

    def _charToIndex(self,ch):
        return ord(ch)-ord('a')

    def insert(self,key):
        pCrawl = self.root
        length = len(key)
        for level in range(length):
            index = self._charToIndex(key[level])

            if not pCrawl.children[index]:
                pCrawl.children[index] = self.getNode()
            pCrawl = pCrawl.children[index]

        pCrawl.isEndOfWord = True

    def search(self, key):
        pCrawl = self.root
        length = len(key)
        for level in range(length):
            index = self._charToIndex(key[level])
            if not pCrawl.children[index]:
                return False
            pCrawl = pCrawl.children[index]

        return pCrawl != None and pCrawl.isEndOfWord


def main():


    keys = ["cat","car","pick","pickle","dog","do"]
    output = ["Not present","Present"]

    t = Trie()
    for key in keys:
        t.insert(key)

    # Search for different keys
    print("{} ---- {}".format("car",output[t.search("car")]))
    print("{} ---- {}".format("do",output[t.search("do")]))
    print("{} ---- {}".format("picke",output[t.search("picke")]))

if __name__ == '__main__':
    main()